### ISSUE TITLE
Need to clean cache for build to succeed when we should not need to

### ISSUE THREAD

**Issue by jurahul** [bazelbuild/bazel#13135]
This is a continuation of a tensorflow build issues that was recently noticed:

https://github.com/tensorflow/tensorflow/issues/47386

Change https://github.com/tensorflow/tensorflow/commit/d439d8c6d4137e0bc468a2de06e41b375029ee92 in the tensorflow repo added a missing dependency in a BUILD file. That caused tensorflow build failures for several folks (as noted on the issue) and the builds worked fine after deleting the bazel cache.

This seems to point to some stale state that was leftover and not correctly purged by bazel. I'd expect we generally should be able to build without cleaning the cache for a change like that. This bug is to track why deleting cache was needed.

I think this is tensorflow build on Linux.

**Comment by jin** [comment]
Setting this as a P1 because it's reported to be an incrementality bug only fixable through a bazel clean, which shouldn't be happening. @oquenchil, could you take a look given that this looks like it's coming from the cc rules?

**Comment by oquenchil** [comment]
Can this be reproduced?

**Comment by jurahul** [comment]
Yes, please look at the associated tensorflow issue for repro: https://github.com/tensorflow/tensorflow/issues/47386

**Comment by nouiz** [comment]
Any update. We are a way people hitting this.
Now doing bisect with TF get way more complicated.
If you combined with the very long TF compilation time that takes hours on a normal workstations, the development workflow on TF is very ugly right now.

**Comment by jurahul** [comment]
Or as a stopgap if there was a way to do "bazel clean --expunge" but only for a particular set of targets, that might help here as well.

**Comment by nouiz** [comment]
The bazel version in the container is 3.7.2

Here is the full steps to reproduce the problem.
```
docker run --gpus=all -it tensorflow/tensorflow:devel-gpu /bin/bash
cd tensorflow_src/
git checkout a2939994e01e377f6ee9a40fdc49e4e1abb9f80b
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu
git checkout e5bf789
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu
```

Give me this error:
```
ERROR: /tensorflow_src/tensorflow/compiler/xla/service/gpu/BUILD:437:16: undeclared inclusion(s) in rule '//tensorflow/compiler/xla/service/gpu:nccl_collective_thunks':
this rule is missing dependency declarations for the following files included by 'tensorflow/compiler/xla/service/gpu/nccl_all_to_all_thunk.cc':
  'bazel-out/k8-opt/bin/tensorflow/compiler/mlir/hlo/_virtual_includes/lhlo_ops_inc_gen/mlir-hlo/Dialect/mhlo/IR/lhlo_ops_structs.h.inc'
```

But if I empty the cache, all works:
```
bazel clean --expunge
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu
```

Output:
```
[10,527 / 10,527] 2 / 2 tests; checking cached actions; last test: //tensorflow/compiler/xla/tests:conv_depthINFO: 10527 processes: 3632 internal, 6895 local.
[10,527 / 10,527] 2 / 2 tests; checking cached actions; last test: //tensorflow/compiler/xla/tests:conv_depthINFO: Build completed successfully, 10527 total actions
//tensorflow/compiler/xla/tests:conv_depthwise_test_gpu                  PASSED in 72.9s
  Stats over 50 runs: max = 72.9s, min = 2.0s, avg = 42.0s, dev = 32.1s
//tensorflow/compiler/xla/tests:grouped_convolution_test_gpu             PASSED in 72.8s
  Stats over 50 runs: max = 72.8s, min = 2.0s, avg = 34.1s, dev = 34.0s

Executed 2 out of 2 tests: 2 tests pass
```

The TensorFlow commit e5bf789 is one that fix a missing dependency. The commits in between will have a real missing build dependency error that isn't fixed by a `bazel clean --expunge`.

**Comment by pcjanzen** [comment]
Does `--spawn_strategy=sandboxed` help?

**Comment by nouiz** [comment]
I tested the 3 new combination:
--spawn_strategy=sandboxed on first bazel command only. Doesn't fix the problem
--spawn_strategy=sandboxed on the second bazel command only. Fix the problem.
--spawn_strategy=sandboxed on both bazel command. Fix the problem.

So this would be a better WAR then using `bazel clean --expunge` as it doesn't trash the cache.
And now we will be able to make git bisect again without more trouble then before.

Any idea what could cause this bug?

Also the web site seems inconsistent in the documentation of the --spawn_strategy parameter.
This page tell that sandboxed whould be the default when supported:
https://docs.bazel.build/versions/master/user-manual.html#flag--spawn_strategy
But this page tell the default is: "remote,worker,sandboxed,local":
https://docs.bazel.build/versions/master/command-line-reference.html#flag--spawn_strategy

I think the first URL need to be updated.

**Comment by pcjanzen** [comment]
When actions execute on the local host, sandboxed execution is used by default when it's supported. However, if remote execution or persistent workers are supported, that's preferred. 

tensorflow overrides the defaults, which is why you see this problem:
https://github.com/tensorflow/tensorflow/blob/master/.bazelrc#L109

The problem is that bazel doesn't remove the files generated by previous builds from the bazel-out directory, but some compilation action still requires a header from the directory that contained them. So the compiler gets an `-I` path that allows it to find the zombies, instead of the ones that are currently declared in the BUILD files. Bazel correctly determines that the compiler has read the wrong version of the file.

With sandboxing on, Bazel creates symlinks only for the inputs currently declared in BUILD files, so the compiler cannot find the remnants of the previous builds in any include path.

**Comment by nouiz** [comment]
Thanks @pcjanzen for the explanation.

We can't go back and edit old TF repo.
So the only solution left is to remove the TF override to prevent that kind of problem in the future and manually add `--spawn_strategy=sandboxed` when compiling past TF when it is needed?
Do you see other options?

@jurahul what do you think of removing that parameter in TF build? Any idea why it was added?

**Comment by nouiz** [comment]
@jurahul do you think we can update TF build script?

I just got this error again and adding `--spawn_strategy=sandboxed` to the command line didn't fix. A `bazel clean` didn't fix it. But a `bazel clean --expunge` did. Any idea why `--spawn_strategy=sandboxed` didn't work?

**Comment by jurahul** [comment]
Apologies I haven't looked into this yet. I'll start looking this week (to see if we can make --spawn_strategy=sandboxed the default for TF builds).

**Comment by jurahul** [comment]
@nouiz, I have update the TF builds to use the sandboxed strategy by default. LMK if that is working for you and we can close this issue.

**Comment by nouiz** [comment]
I tested like bellow. It work. Thanks for the fix. So we can close this issue.
I still do not know why doing it manually didn't work for me yesterday. But we can forget that and suppose all is good for now.

```
docker run --gpus=all -it tensorflow/tensorflow:devel-gpu /bin/bash
cd tensorflow_src/
git checkout a2939994e01e377f6ee9a40fdc49e4e1abb9f80b
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu

#The commit just before the fix.
git checkout 3a8e7077ced168e8e67f9db065866ebc3cef8054
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu
# It failed to compile as expected

git checkout 9171e39a2a1c4b057639c3d6e7c53db627a461ae
bazel test -c opt --config=cuda //tensorflow/compiler/xla/tests:conv_depthwise_test_gpu //tensorflow/compiler/xla/tests:grouped_convolution_test_gpu
# No crash!
```

**Comment by jurahul** [comment]
Thanks for confirming. Closing this (and the TF issue that spawned this is closed already) based on @nouiz's comment above.

**Comment by jurahul** [comment]
@nouiz, I had to revert the fix as it caused a windows build failure. Working on a new one.

**Comment by nouiz** [comment]
Thanks for the update.

**Comment by nouiz** [comment]
@jurahul any update? Your first fix broke windows. What about doing it only on linux? That would be good for me.

**Comment by jurahul** [comment]
Yes, unfortunately my second attempt to fix this by just enabling for Linux did not go well as there were more build failures seen internally. So unfortunately this is not a straightforward fix it seems. I'll have to spend more time here to see what the issues were.

**Comment by matthauck** [comment]
Running into this as well. Some BUILD file changes seem to require `bazel clean --expunge` before the build works again. (This is obviously not a trivial thing to do for all CI nodes)

- We do override spawn_strategy on linux since we noticed `sandboxed` can be very slow inside a docker container
- We do not override spawn_strategy on windows. 

We have witnessed this on both windows and linux, and honestly it is detracting from our bazel transition in questioning the correctness of bazel incremental builds. 

Not quite sure why this issue was closed -- this is a legitimate bug in bazel, isn't it?
